Update Locations Pages to Use a Persisted Dataset

richarddubay opened the pull request:
## DESCRIPTION

### What does this PR do, or why is it needed?
This PR: 
* Updates location lava templates to pull their data from a persisted dataset.
* Changes the "Get Directions" buttons to be regular links. 

### How do I test this PR?
* Copy the persisted dataset from https://rich-rock.newspring.cc/page/2396?PersistedDatasetId=1 and create or update the Campus dataset in your database.
* Pull this code and then run your site.
* Look at the locations pages ( `/locations`, `/locations/{campus}`, `/fuse/locations`, and `/fuse/locations/{campus}` and make sure that everything loads and looks correct.

## TODO

- [x] I am affirming this is my _best_ work ([Ecclesiastes 9:10](https://www.bible.com/bible/97/ECC.9.10.MSG))
- [x] PR has a relevant title that will be understandable in a public changelog (ie...non developers)
- [x] Testing info includes any items that need to be added to a local Rock instance in order to test and/or the database against which it can be tested.
- [x] Upload GIF(s) of relevant changes
- [x] Set a relevant reviewer

## REVIEW

- [ ] Review code through the lens of being concise, simple, and well-documented
- [ ] Manual QA to ensure the changes look/behave as expected

> The purpose of PR Review is to _improve the quality of the software._

briankalwat commented:
@richarddubay  this is the lava that generates the dataset on your dev server:
`[
    {% campus where:'IsActive == "True" && Public == "Yes"' sort:'Name' iterator:'campuses' %}
    {%- for campus in campuses -%}
        {% assign campusType = campus | Attribute:'CampusType','RawValue' %}
        {% if campusType == 'permanent' or campusType == 'portable' or campusType == 'community' %}
            {% group where:'ParentGroupId == 1528254 && Name == "{{ campus.Name }}"' iterator:'groups' %}
                {% for group in groups %}
                    {% assign campusStaff =  group.Members | Select:'Person' | Sort:'LastName' %}
                {% endfor %}
            {% endgroup %}
            {
                "Id": {{ campus.Id | ToJSON }},
                "Guid": {{ campus.Guid | ToJSON }},
                "Shortcode": {{ campus.ShortCode | ToJSON }},
                "CampusType": {{ campusType | ToJSON }},
                "Name": {{ campus.Name | ToJSON }},
                "LocationStreet1": {{ campus.Location.Street1 | ToJSON }},
                "LocationStreet2": {{ campus.Location.Street2 | ToJSON }},
                "LocationCity": {{ campus.Location.City | ToJSON }},
                "LocationState": {{ campus.Location.State | ToJSON }},
                "LocationPostalCode": {{ campus.Location.PostalCode | ToJSON }},
                "PhoneNumber": {{ campus.PhoneNumber | ToJSON }},
                "Email": {{ campus.Name | Replace:' ','' | Downcase | Append:'@newspring.cc' | ToJSON }},
                "Url": {{ campus.Name | Replace:' ','' | Downcase | Prepend:'https://newspring.cc/locations/' | ToJSON }},
                "CampusPastorPersonId": {{ campus.LeaderPersonAliasId | PersonByAliasId | Property:'Id' | ToJSON }},
                "FusePastorPersonId": {{ campus | Attribute:'FusePastor','Id' | ToJSON }},
                "GooglePlacesId": {{ campus | Attribute:'GooglePlacesId' | ToJSON }},
                "GetDirectionsURL": {{ campus | Attribute:'GetDirectionsURL','RawValue' | ToJSON }},
                "FuseGetDirectionsURL": {{ campus | Attribute:'FuseGetDirectionsURL','RawValue' | ToJSON }},
                "LandscapeImage": {{ campus | Attribute:'ImageLandscape' | ToJSON }},
                "FuseLandscapeImage": {{ campus | Attribute:'FuseImageLandscape' | ToJSON }},
                "ServiceTimes": {{ campus | Attribute:'ServiceTimes1' | FromJSON | ToJSON }},
                "MailingAddress": {{ campus | Attribute:'MailingAddress' | ToJSON }},
                "OfficeAddress": {{ campus | Attribute:'OfficeAddress' | ToJSON }},
                "CampusStaff": [
                    {% for staffMember in campusStaff %}
                        {
                            "PersonId": {{ staffMember.Id }},
                            "FullName": {{ staffMember.FullName | ToJSON }},
                            "StaffImage": {{ staffMember | Attribute:'StaffImage','Url' | ToJSON }},
                            "StaffTitle": {{ staffMember | Attribute:'StaffTitle' | ToJSON }},
                            "StaffEmail": {{ staffMember | Attribute:'StaffEmail' | ToJSON }},
                            "NickName": {{ staffMember.NickName | ToJSON }}
                        }{%- unless forloop.last -%},{%- endunless -%}
                    {% endfor %}
                ]
            }{%- unless forloop.last -%},{%- endunless -%}
        {% endif %}
    {%- endfor -%}
    {% endcampus %}
]`

The recommendation I have revolves around changing the way we're getting campus staff. Currently, we calling the db to get the staff group for every individual campus, a total of 14 separate calls. What I think we should do is make one call to get all of the staff groups, create a lava object out of that, and then access that object from inside of the campus loop. Here's what I was thinking - I'm testing it with the other lava changes you made to make sure that it works the same way.

`{% capture staffGroupsString %}
[{% group where:'ParentGroupId == 1528254' iterator:'groups' %}
    {% for group in groups %}{
        "Id":{{ group.Id }},
        "Name":{{ group.Name | ToJSON }},
        "Staff":{{ group.Members | Select:'Person' | Sort:'LastName' | Select:'Id' | ToJSON }}
    }{% unless forloop.last %},{% endunless %}{% endfor %}
{% endgroup %}]
{% endcapture %}
{% assign staffGroups = staffGroupsString | FromJSON %}

[
    {% campus where:'IsActive == "True" && Public == "Yes"' sort:'Name' iterator:'campuses' %}
    {%- for campus in campuses -%}
        {% assign campusType = campus | Attribute:'CampusType','RawValue' %}
        {% if campusType == 'permanent' or campusType == 'portable' or campusType == 'community' %}
            {% assign staffMembers = staffGroups | | Where:'Name', 'Clemson' | First | Property:'Staff' %}
            {
                "Id": {{ campus.Id | ToJSON }},
                "Guid": {{ campus.Guid | ToJSON }},
                "Shortcode": {{ campus.ShortCode | ToJSON }},
                "CampusType": {{ campusType | ToJSON }},
                "Name": {{ campus.Name | ToJSON }},
                "LocationStreet1": {{ campus.Location.Street1 | ToJSON }},
                "LocationStreet2": {{ campus.Location.Street2 | ToJSON }},
                "LocationCity": {{ campus.Location.City | ToJSON }},
                "LocationState": {{ campus.Location.State | ToJSON }},
                "LocationPostalCode": {{ campus.Location.PostalCode | ToJSON }},
                "PhoneNumber": {{ campus.PhoneNumber | ToJSON }},
                "Email": {{ campus.Name | Replace:' ','' | Downcase | Append:'@newspring.cc' | ToJSON }},
                "Url": {{ campus.Name | Replace:' ','' | Downcase | Prepend:'https://newspring.cc/locations/' | ToJSON }},
                "CampusPastorPersonId": {{ campus.LeaderPersonAliasId | PersonByAliasId | Property:'Id' | ToJSON }},
                "FusePastorPersonId": {{ campus | Attribute:'FusePastor','Id' | ToJSON }},
                "GooglePlacesId": {{ campus | Attribute:'GooglePlacesId' | ToJSON }},
                "GetDirectionsURL": {{ campus | Attribute:'GetDirectionsURL','RawValue' | ToJSON }},
                "FuseGetDirectionsURL": {{ campus | Attribute:'FuseGetDirectionsURL','RawValue' | ToJSON }},
                "LandscapeImage": {{ campus | Attribute:'ImageLandscape' | ToJSON }},
                "FuseLandscapeImage": {{ campus | Attribute:'FuseImageLandscape' | ToJSON }},
                "ServiceTimes": {{ campus | Attribute:'ServiceTimes1' | FromJSON | ToJSON }},
                "MailingAddress": {{ campus | Attribute:'MailingAddress' | ToJSON }},
                "OfficeAddress": {{ campus | Attribute:'OfficeAddress' | ToJSON }},
                "CampusStaff": [
                    {% for staff in staffMembers %}
                        {% assign staffMember = staff | PersonById %}
                        {
                            "PersonId": {{ staffMember.Id }},
                            "FullName": {{ staffMember.FullName | ToJSON }},
                            "StaffImage": {{ staffMember | Attribute:'StaffImage','Url' | ToJSON }},
                            "StaffTitle": {{ staffMember | Attribute:'StaffTitle' | ToJSON }},
                            "StaffEmail": {{ staffMember | Attribute:'StaffEmail' | ToJSON }},
                            "NickName": {{ staffMember.NickName | ToJSON }}
                        }{%- unless forloop.last -%},{%- endunless -%}
                    {% endfor %}
                ]
            }{%- unless forloop.last -%},{%- endunless -%}
        {% endif %}
    {%- endfor -%}
    {% endcampus %}
]`

briankalwat commented on RockWeb/Themes/NewSpring/assets/Lava/UI/locations-hero.lava:
```diff
@@ -1,44 +1,43 @@
-{% for campus in campuses %}
+{% assign campusSlug = 'Global' | PageParameter:'CampusSlug' | Replace:'-',' ' | Capitalize %}
+{% assign campuses = 'campuses' | PersistedDataset %}
+{% assign campus = campuses | Where:'Name', campusSlug | First %}
```
You can actually combine lines 1 & 2:
`{% assign campus = 'campuses' | PersistedDataset | Where:'Name', campusSlug | First %}`

briankalwat commented on RockWeb/Themes/NewSpring/assets/Lava/UI/locations-hero.lava:
```diff
@@ -1,44 +1,43 @@
+{% endcomment %}
+{% capture title %}{% if servicetype == 'Fuse' %}Fuse Student Ministry{% else %}{{ 'Global' | Attribute:'OrganizationName' }}{% endif %}{% if campus.Name != 'Eastlan' %} in {{ campus.Name }}, SC{% else %} - {{ campus.Name }}{% endif %} | Locations{% if servicetype == 'Fuse' %} | Fuse{% endif %} | {{ 'Global' | Attribute:'OrganizationName' }}{% endcapture %}
+{{ title | SetPageTitle:'PageTitle' }}
+{{ title | SetPageTitle:'BrowserTitle' }}
```
Probably don't need both of these, if you leave off the parameter, it sets both I believe. Did we land on leaving this here and removing it from the block?
`{{ title | SetPageTitle }}`

richarddubay replied:
I don't think we ever talked about this at all. This part of the code actually didn't change. But I can for sure change it here if you think it makes sense to do so.

briankalwat replied:
I thought this was talking about this code? https://newspring.slack.com/archives/D04FDSWPC/p1585229590017800

richarddubay replied:
OH YEAH! I totally forgot about that. 🤦‍♂️

Yeah, I think we landed on putting this in the lava template and removing it from the block (because it wasn't really doing anything in the block anyway).

briankalwat commented on RockWeb/Themes/NewSpring/assets/Lava/UI/locations-information.lava:
```diff
@@ -1,11 +1,9 @@
-{% assign getDirections = campus | Attribute:'GetDirectionsURL','RawValue' %}
+{% assign campusSlug = 'Global' | PageParameter:'CampusSlug' | Replace:'-',' ' | Capitalize %}
+{% assign campuses = 'campuses' | PersistedDataset %}
+{% assign campus = campuses | Where:'Name', campusSlug | First %}
```
Same here, we can combine 2 + 3 by just continuing to add the filters.
`{% assign campus = 'campuses' | PersistedDataset | Where:'Name', campusSlug | First %}`

briankalwat commented on RockWeb/Themes/NewSpring/assets/Lava/UI/locations-leader.lava:
```diff
@@ -1,54 +1,45 @@
-    {[ section title:'{{ campus.Name }} Staff' ]}
+{% assign campusSlug = 'Global' | PageParameter:'CampusSlug' | Replace:'-',' ' | Capitalize %}
+{% assign campuses = 'campuses' | PersistedDataset %}
+{% assign campus = campuses | Where:'Name', campusSlug | First %}
```
Combine lines 2 + 3:
`{% assign campus = 'campuses' | PersistedDataset | Where:'Name', campusSlug | First %}`

briankalwat commented on RockWeb/Themes/NewSpring/assets/Lava/UI/locations-swiper.lava:
```diff
@@ -2,25 +2,20 @@
+    {% for campus in campuses %}
+      {% assign directionsURL = campus.GetDirectionsURL %}
+      {% assign fuseDirectionsUrl = campus.FuseGetDirectionsURL %}
+      [[ item data:'data-latlon="{{ campus.Location.Latitude }},{{ campus.Location.Longitude }}"' ]]
```
I don't think this is working because I don't see Latitude/Longitude in the dataset.

richarddubay commented:
@briankalwat Okay, I updated based on your suggestions. I added latitude and longitude to the dataset as well. The search in the hero isn't working though. Any thoughts?

briankalwat commented:
@richarddubay it's working on my dev environment? Are you talking about the hero on /locations? Do you get any console errors when you submit your zip code or anything?

richarddubay commented:
@briankalwat Nope, no errors ... but it's not doing any sorting either. Perhaps I'm just missing something else somewhere? I tried yours and it seems to work but it scares me that I don't know why mine won't.

briankalwat commented:
Probably missing a script or something, I can look into it.

On Tue, Mar 31, 2020 at 2:56 PM Richard Dubay <notifications@github.com>
wrote:

> @briankalwat <https://github.com/briankalwat> Nope, no errors ... but
> it's not doing any sorting either. Perhaps I'm just missing something else
> somewhere? I tried yours and it seems to work but it scares me that I don't
> know why mine won't.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/NewSpring/Rock/pull/1067#issuecomment-606810458>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAS2AH5P6IUWZOAB6TVBIDDRKI4FRANCNFSM4LUM2IKQ>
> .
>


-- 
*Brian Kalwat*
Software Engineer
NewSpring Church
newspring.cc <http://www.newspring.cc>
(864) 442-2288

richarddubay commented:
Thanks. I'd just feel better knowing it actually worked on mine. But you're good if I merge this?
